src/util/syscall_sandbox.cpp

```diff
@@ -0,0 +1,450 @@
+    switch (syscall_policy) {
+    case SyscallSandboxPolicy::INITIALIZATION: // Thread: main thread (state: init)
+        // SyscallSandboxPolicy::INITIALIZATION is the first policy loaded.
+        //
+        // Subsequently loaded policies can reduce the abilities further, but
+        // abilities can never be regained.
+        //
+        // SyscallSandboxPolicy::INITIALIZATION must thus be a superset of all
+        // other policies.
+        seccomp_policy_builder.AllowFileSystem();
+        seccomp_policy_builder.AllowNetwork();
+        break;
+    case SyscallSandboxPolicy::INITIALIZATION_DNS_SEED: // Thread: dnsseed
+        seccomp_policy_builder.AllowFileSystem();
+        seccomp_policy_builder.AllowNetwork();
+        break;
+    case SyscallSandboxPolicy::INITIALIZATION_LOAD_BLOCKS: // Thread: loadblk
+        seccomp_policy_builder.AllowFileSystem();
+        break;
+    case SyscallSandboxPolicy::INITIALIZATION_TOR_CONTROL: // Thread: torcontrol
+        seccomp_policy_builder.AllowFileSystem();
+        seccomp_policy_builder.AllowNetwork();
+        break;
+    case SyscallSandboxPolicy::INITIALIZATION_UPNP: // Thread: upnp
+        seccomp_policy_builder.AllowNetwork();
+        break;
+    case SyscallSandboxPolicy::MESSAGE_HANDLER: // Thread: msghand
+        seccomp_policy_builder.AllowFileSystem();
+        break;
+    case SyscallSandboxPolicy::NET: // Thread: net
+        seccomp_policy_builder.AllowFileSystem();
+        seccomp_policy_builder.AllowNetwork();
+        break;
+    case SyscallSandboxPolicy::NET_ADD_CONNECTION: // Thread: addcon
+        seccomp_policy_builder.AllowFileSystem();
+        seccomp_policy_builder.AllowNetwork();
+        break;
+    case SyscallSandboxPolicy::NET_HTTP_SERVER: // Thread: http
+        seccomp_policy_builder.AllowFileSystem();
+        seccomp_policy_builder.AllowNetwork();
+        break;
+    case SyscallSandboxPolicy::NET_HTTP_SERVER_WORKER: // Thread: httpworker.<N>
+        seccomp_policy_builder.AllowFileSystem();
+        seccomp_policy_builder.AllowNetwork();
+        break;
+    case SyscallSandboxPolicy::NET_OPEN_CONNECTION: // Thread: opencon
+        seccomp_policy_builder.AllowNetwork();
+        break;
+    case SyscallSandboxPolicy::SCHEDULER: // Thread: scheduler
+        seccomp_policy_builder.AllowFileSystem();
+        break;
+    case SyscallSandboxPolicy::TX_INDEX: // Thread: txindex
+        seccomp_policy_builder.AllowFileSystem();
+        break;
+    case SyscallSandboxPolicy::VALIDATION_SCRIPT_CHECK: // Thread: scriptch.<N>
+        break;
+    case SyscallSandboxPolicy::SHUTOFF: // Thread: main thread (state: shutoff)
+        seccomp_policy_builder.AllowFileSystem();
+        break;
+    }
+
+    const SyscallSandboxDefaultAction default_action = std::getenv("EXPERIMENTAL_SYSCALL_SANDBOX_MODE_KILL_THREAD") != nullptr ? SyscallSandboxDefaultAction::KILL_THREAD : SyscallSandboxDefaultAction::DEBUG_SIGNAL_HANDLER;
```

Comment:
Another self-review:

Instead of opt-ing out of debug mode via an environment variable we should make it the other way around: the debug mode should be opt-in via `--enable-debug`.

In other words: `SECCOMP_RET_KILL_PROCESS` should be the default action, and `SECCOMP_RET_TRAP` should only be used if compiled with `--enable-debug`.

Background:

When using `SECCOMP_RET_KILL_PROCESS` the kernel immediately kills off the offending process in case of a syscall violation.

When using `SECCOMP_RET_TRAP` the kernel sends a `SIGSYS` signal to the offending process in case of a syscall violation: that allows us to print a user-friendly error message via a signal handler. Note that the attacker may be able to control the execution in the signal handler (although still with restricted syscall access), so this mode should really only be used when debugging.

Reply:
Not sure on this. I do think feedback to the user is very important. I am slightly worried when if our syscall profile isn't 100% up to date (my experience is that this can *easily* happen especially with dynamically linked glibc), something starts killing the process without any means to diagnose or debug, that's extremely frustrating.

Reply:
Yes, I guess this is the classical trade-off between security vs debuggability/ease-of-use :)

Making `SECCOMP_RET_KILL_PROCESS` the default is slightly better from a security perspective (no risk that the attacker is getting a "second chance" via the signal handler) whereas making `SECCOMP_RET_TRAP` the default is slightly better from a debuggability/ease-of-use perspective.

I'll let others chime in here: I'll happily adjust to the consensus opinion :)

The following shows the differences in output between the two modes of operation in case of a simulated syscall violation.

When using `./configure --enable-syscall-sandbox`:

```
$ make distclean
$ ./autogen.sh
$ ./configure --enable-syscall-sandbox
$ make
$ src/bitcoind
2021-03-16T12:34:56Z Bitcoin Core version v21.99.0-[…]-dirty (release build)
…
Bad system call (core dumped)
```

When using `./configure --enable-syscall-sandbox --enable-debug`:

```
$ make distclean
$ ./autogen.sh
$ ./configure --enable-syscall-sandbox --enable-debug
$ make
$ src/bitcoind
2021-03-16T12:34:56Z Bitcoin Core version v21.99.0-[…]-dirty (debug build)
…
ERROR: The syscall "execve" (syscall number 59) is not allowed by the syscall sandbox in thread "msghand". Please report. Exiting.
2021-03-16T12:34:56Z ERROR: The syscall "execve" (syscall number 59) is not allowed by the syscall sandbox in thread "msghand". Please report. Exiting.
terminate called without an active exception
Aborted (core dumped)
```

Reply:
@laanwj Now updated to default to debug mode as you suggested :)

Updated OP to describe the two modes of operation:
* Debug mode (enabled via `configure` flag `--with-syscall-sandbox`): If a non-allowlisted syscall is called a debug handler will be invoked. Debug info will be printed (`ERROR: The syscall "execve" (syscall number 59) is not allowed by the syscall sandbox in thread "msghand". Please report. Exiting.`) and `std::terminate` is called. Note that the attacker may be able to control the execution in the signal handler (although still with restricted syscall access), so this mode should really only be used when debugging.
* Kill without debug mode (enabled via `configure` flag `--with-syscall-sandbox=kill-without-debug`): If a non-allowlisted syscall is called the kernel will immediately kill the offending process.